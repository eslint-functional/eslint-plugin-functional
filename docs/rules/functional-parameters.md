# Enforce Functional Parameters (functional-parameters)

Disallow use of rest parameters, the `arguments` keyword and enforces that functions take a least 1 parameter.

## Rule Details

In functions, `arguments` is a special variable that is implicitly available.
This variable is an array containing the arguments passed to the function call and is often used to allow for any number of parameters to be passed to a function. Rest parameters are another way any number of parameters can be passed to a function.

When it comes to functional programming, known and explicit parameters must be used.

Note: With an unknown number of parameters, currying functions is a lot more difficult/impossible.

Examples of **incorrect** code for this rule:

```js
/* eslint functional/functional-parameters: "error" */

function add() {
  return arguments.reduce((sum, number) => sum + number, 0);
}
```

```js
/* eslint functional/functional-parameters: "error" */

function add(...numbers) {
  return numbers.reduce((sum, number) => sum + number, 0);
}
```

Examples of **correct** code for this rule:

```js
/* eslint functional/functional-parameters: "error" */

function add(numbers) {
  return numbers.reduce((sum, number) => sum + number, 0);
}
```

## Options

This rule accepts an options object of the following type:

```ts
{
  allowRestParameter: boolean;
  allowArgumentsKeyword: boolean;
  enforceParameterCount: false | "atLeastOne" | "exactlyOne" | {
    count: "atLeastOne" | "exactlyOne";
    ignoreIIFE: boolean;
  };
  ignorePattern?: string | Array<string>;
}
```

The default options:

```ts
{
  allowRestParameter: false,
  allowArgumentsKeyword: false,
  enforceParameterCount: {
    count: "atLeastOne",
    ignoreIIFE: true
  }
}
```

Note: the `lite` ruleset overrides the default options to:

```ts
{
  allowRestParameter: false,
  allowArgumentsKeyword: false,
  enforceParameterCount: false
}
```

### `allowRestParameter`

If true, this option allows for the use of rest parameters.

### `allowArgumentsKeyword`

If true, this option allows for the use of the `arguments` keyword.

### `enforceParameterCount`

This option enforces the number of parameters each function takes.

#### false

Don't enforce a parameter count.

#### "atLeastOne"

Require all functions to have a least one parameter.

There's not much point of having a function that doesn't take any parameters in functional programming.

#### "exactlyOne"

Require all functions to have exactly one parameter.

Any function that take takes multiple parameter can be rewritten as a higher-order function that only takes one.

Example:

```js
// This function
function add(x, y) {
  return x + y;
}

// can be rewritten like this.
function add(x) {
  return (y) => x + y;
}
```

See [Currying](https://en.wikipedia.org/wiki/Currying) and [Higher-order function](https://en.wikipedia.org/wiki/Higher-order_function) on Wikipedia for more infomation.

#### `enforceParameterCount.count`

See [enforceParameterCount](#enforceparametercount).

#### `enforceParameterCount.ignoreIIFE`

If true, this option allows for the use of [IIFEs](https://developer.mozilla.org/en-US/docs/Glossary/IIFE) that do not have any parameters.

### `ignorePattern`

Patterns will be matched against function names.
See the [ignorePattern](./options/ignore-pattern.md) docs for more infomation.
