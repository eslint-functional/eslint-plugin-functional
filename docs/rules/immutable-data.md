# Disallow mutating objects and arrays (immutable-data)

This rule prohibits syntax that mutates existing objects and arrays via assignment to or deletion of their properties/elements.

## Rule Details

While requiring the `readonly` modifier forces declared types to be immutable,
it won't stop assignment into or modification of untyped objects or external types declared under different rules.

Examples of **incorrect** code for this rule:

```js
/* eslint functional/immutable-data: "error" */

const obj = { foo: 1 };

obj.foo += 2; // <- Modifying an existing object/array is not allowed.
obj.bar = 1; // <- Modifying an existing object/array is not allowed.
delete obj.foo; // <- Modifying an existing object/array is not allowed.
Object.assign(obj, { bar: 2 }); // <- Modifying properties of existing object not allowed.
```

```js
/* eslint functional/immutable-data: "error" */

const arr = [0, 1, 2];

arr[0] = 4; // <- Modifying an array is not allowed.
arr.length = 1; // <- Modifying an array is not allowed.
arr.push(3); // <- Modifying an array is not allowed.
```

Examples of **correct** code for this rule:

```js
/* eslint functional/immutable-data: "error" */

const obj = { foo: 1 };
const arr = [0, 1, 2];

const x = {
  ...obj
  bar: [
    ...arr, 3, 4
  ]
}
```

## Options

This rule accepts an options object of the following type:

```ts
{
  assumeTypes:
    | boolean
    | {
        forArrays: boolean;
        forObjects: boolean;
      }
  ignoreImmediateMutation: boolean;
  ignorePattern?: string | Array<string>;
  ignoreAccessorPattern?: string | Array<string>;
}
```

The default options:

```ts
{
  assumeTypes: true,
  ignoreImmediateMutation: true,
};
```

### `assumeTypes`

The rule take advantage of TypeScript's typing engine to check if mutation is taking place.
If you are not using TypeScript, type checking cannot be performed; hence this option exists.

This option will make the rule assume the type of the nodes it is checking are of type Array/Object.  
However this may result in some false positives being picked up.

Disabling this option can result in false negatives, for example:

```js
// When this option is DISABLED (and type info is not available).
const x = [0, 1, 2];
x.push(3); // This will NOT be flagged.
// This is due to the fact that without a typing engine, we cannot tell that x is an array.
```

Note: This option will have no effect if the TypeScript typing engine is avaliable (i.e. you are using TypeScript and have configured eslint correctly).

### `ignoreImmediateMutation`

If true, immediate mutation of objects before they are assigned to a variable is allowed.
This allows for the use of array mutator methods to be chained to newly created arrays.

For example, an array can be immutably sorted like so:

```js
const original = ["foo", "bar", "baz"];
const sorted = [...original].sort((a, b) => a.localeCompare(b)); // This is OK with ignoreImmediateMutation.
```

### `ignorePattern`

Patterns will be matched against variable names.
See the [ignorePattern](./options/ignore-pattern.md) docs for more infomation.

### `ignoreAccessorPattern`

This option takes a match string or an array of match strings (not a RegExp pattern).

The match string allows you to specify dot seperated `.` object paths and has support for "glob" `*` and "globstar" `**` matching.

For example:

```js
{
  // Ignore all reassigning to object properties that are prefixed with "mutable_".
  "ignoreAccessorPattern": "**.mutable_*"
}
```

```js
{
  // Ignore all shallow mutations made to object properties that are prefixed with "mutable_".
  "ignoreAccessorPattern": "**.mutable_*.*"
}
```

```js
{
  // Ignore all deep mutations made to object properties that are prefixed with "mutable_".
  "ignoreAccessorPattern": "**.mutable_*.*.**"
}
```

```js
{
  // Ignore all deep mutations and reassigning to object properties that are prefixed with "mutable_".
  "ignoreAccessorPattern": "**.mutable_*.**"
  // This is the same as `"ignoreAccessorPattern": ["**.mutable_*", "**.mutable_*.*.**"]`
}
```

#### Wildcards

The following wildcards can be used when specifing a pattern:

`**` - Match any depth (including zero). Can only be used as a full accessor.
`*` - When used as a full accessor, match the next accessor (there must be one). When used as part of an accessor, match any characters.
