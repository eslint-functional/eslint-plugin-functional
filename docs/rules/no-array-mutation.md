### no-array-mutation

[![Type Info Required][type-info-badge]][type-info-url]

This rule prohibits mutating an array via assignment to or deletion of their elements/properties. This rule enforces array immutability without the use of `ReadonlyArray<T>` (as apposed to [readonly-array](#readonly-array)).

```typescript
const x = [0, 1, 2];

x[0] = 4; // <- Mutating an array is not allowed.
x.length = 1; // <- Mutating an array is not allowed.
x.push(3); // <- Mutating an array is not allowed.
```

#### Has Fixer

No

#### Options

- [ignore-prefix](#using-the-ignore-prefix-option)
- [ignore-suffix](#using-the-ignore-suffix-option)
- [ignore-pattern](#using-the-ignore-pattern-option)
- [ignore-new-array](#using-the-ignore-new-array-option-with-no-array-mutation)
- ~~ignore-mutation-following-accessor~~ - _deprecated in favor of [ignore-new-array](#using-the-ignore-new-array-option-with-no-array-mutation)_

#### Example config

```javascript
"no-array-mutation": true
```

```javascript
"no-array-mutation": [true, {"ignore-prefix": "mutable"}]
```

```javascript
"no-array-mutation": [true, "ignore-new-array"]
```
