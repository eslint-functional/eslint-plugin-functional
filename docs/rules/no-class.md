# Disallow classes (no-class)

Disallow use of the `class` keyword.

## Rule Details

Thanks to libraries like [recompose](https://github.com/acdlite/recompose) and Redux's [React Container components](http://redux.js.org/docs/basics/UsageWithReact.html), there's not much reason to build Components using `React.createClass` or ES6 classes anymore. The `no-this` rule makes this explicit.

```typescript
const Message = React.createClass({
  render: function() {
    return <div>{this.props.message}</div>; // <- no this allowed
  }
});
```

Instead of creating classes, you should use React 0.14's [Stateless Functional Components](https://medium.com/@joshblack/stateless-components-in-react-0-14-f9798f8b992d#.t5z2fdit6) and save yourself some keystrokes:

```typescript
const Message = ({ message }) => <div>{message}</div>;
```

What about lifecycle methods like `shouldComponentUpdate`? We can use the [recompose](https://github.com/acdlite/recompose) library to apply these optimizations to your Stateless Functional Components. The [recompose](https://github.com/acdlite/recompose) library relies on the fact that your Redux state is immutable to efficiently implement shouldComponentUpdate for you.

```typescript
import { pure, onlyUpdateForKeys } from "recompose";

const Message = ({ message }) => <div>{message}</div>;

// Optimized version of same component, using shallow comparison of props
// Same effect as React's PureRenderMixin
const OptimizedMessage = pure(Message);

// Even more optimized: only updates if specific prop keys have changed
const HyperOptimizedMessage = onlyUpdateForKeys(["message"], Message);
```

## Options

The rule does not accept any options.
