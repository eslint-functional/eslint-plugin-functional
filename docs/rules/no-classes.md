<!-- markdownlint-disable -->
<!-- begin auto-generated rule header -->

# Disallow classes (`functional/no-classes`)

üíº This rule is enabled in the following configs: ‚òëÔ∏è `lite`, `noOtherParadigms`, ‚úÖ `recommended`, üîí `strict`.

<!-- end auto-generated rule header -->
<!-- markdownlint-restore -->
<!-- markdownlint-restore -->

Disallow use of the `class` keyword.

## Rule Details

### ‚ùå Incorrect

<!-- eslint-skip -->

```js
/* eslint functional/no-classes: "error" */

class Dog {
  constructor(name, age) {
    this.name = name;
    this.age = age;
  }

  get ageInDogYears() {
    return 7 * this.age;
  }
}

const dogA = new Dog("Jasper", 2);

console.log(`${dogA.name} is ${dogA.ageInDogYears} in dog years.`);
```

### ‚úÖ Correct

```js
/* eslint functional/no-classes: "error" */

function getAgeInDogYears(age) {
  return 7 * age;
}

const dogA = {
  name: "Jasper",
  age: 2,
};

console.log(`${dogA.name} is ${getAgeInDogYears(dogA.age)} in dog years.`);
```

### React Examples

Thanks to libraries like [recompose](https://github.com/acdlite/recompose) and Redux's
[React Container components](http://redux.js.org/docs/basics/UsageWithReact.html), there's not much reason to build
Components using `React.createClass` or ES6 classes anymore. The `no-this-expressions` rule makes this explicit.

```js
const Message = React.createClass({
  render() {
    return <div>{this.props.message}</div>; // <- no this allowed
  },
});
```

Instead of creating classes, you should use React 0.14's
[Stateless Functional
Components](https://medium.com/@joshblack/stateless-components-in-react-0-14-f9798f8b992d#.t5z2fdit6)
and save yourself some keystrokes:

```js
const Message = ({ message }) => <div>{message}</div>;
```

What about lifecycle methods like `shouldComponentUpdate`?
We can use the [recompose](https://github.com/acdlite/recompose) library to apply these optimizations to your
Stateless Functional Components. The [recompose](https://github.com/acdlite/recompose) library relies on the fact that
your Redux state is immutable to efficiently implement `shouldComponentUpdate` for you.

```js
import { onlyUpdateForKeys, pure } from "recompose";

const Message = ({ message }) => <div>{message}</div>;

// Optimized version of same component, using shallow comparison of props
// Same effect as React's PureRenderMixin
const OptimizedMessage = pure(Message);

// Even more optimized: only updates if specific prop keys have changed
const HyperOptimizedMessage = onlyUpdateForKeys(["message"], Message);
```
