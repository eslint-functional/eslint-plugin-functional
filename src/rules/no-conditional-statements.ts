import type { TSESLint, TSESTree } from "@typescript-eslint/utils";
import type { JSONSchema4 } from "json-schema";
import type { Type } from "typescript";

import tsApiUtils from "~/conditional-imports/ts-api-utils";
import type { RuleResult, NamedCreateRuleMetaWithCategory } from "~/utils/rule";
import { createRule, getTypeOfNode } from "~/utils/rule";
import {
  isBlockStatement,
  isBreakStatement,
  isContinueStatement,
  isExpressionStatement,
  isIfStatement,
  isReturnStatement,
  isSwitchStatement,
  isThrowStatement,
} from "~/utils/type-guards";

/**
 * The name of this rule.
 */
export const name = "no-conditional-statements" as const;

/**
 * The options this rule can take.
 */
type Options = [
  {
    allowReturningBranches: boolean | "ifExhaustive";
  }
];

/**
 * The schema for the rule options.
 */
const schema: JSONSchema4 = [
  {
    type: "object",
    properties: {
      allowReturningBranches: {
        oneOf: [
          {
            type: "boolean",
          },
          {
            type: "string",
            enum: ["ifExhaustive"],
          },
        ],
      },
    },
    additionalProperties: false,
  },
];

/**
 * The default options for the rule.
 */
const defaultOptions: Options = [{ allowReturningBranches: false }];

/**
 * The possible error messages.
 */
const errorMessages = {
  incompleteBranch:
    "Incomplete branch, every branch in a conditional statement must contain a return statement.",
  incompleteIf:
    "Incomplete if, it must have an else statement and every branch must contain a return statement.",
  incompleteSwitch:
    "Incomplete switch, it must be exhaustive or have an default case and every case must contain a return statement.",
  unexpectedIf:
    "Unexpected if, use a conditional expression (ternary operator) instead.",
  unexpectedSwitch:
    "Unexpected switch, use a conditional expression (ternary operator) instead.",
} as const;

/**
 * The meta data for this rule.
 */
const meta: NamedCreateRuleMetaWithCategory<keyof typeof errorMessages> = {
  type: "suggestion",
  docs: {
    category: "No Statements",
    description: "Disallow conditional statements.",
    recommended: "error",
  },
  messages: errorMessages,
  schema,
};

/**
 * Report the given node as an incomplete branch violation.
 *
 * @param node - The node to report.
 * @returns A violation rule result.
 */
function incompleteBranchViolation(
  node: TSESTree.Node
): RuleResult<keyof typeof errorMessages, Options>["descriptors"] {
  return [{ node, messageId: "incompleteBranch" }];
}

/**
 * Get a function that tests if the given statement is never returning.
 */
function getIsNeverExpressions(
  context: TSESLint.RuleContext<keyof typeof errorMessages, Options>
) {
  return (statement: TSESTree.Statement) => {
    if (isExpressionStatement(statement)) {
      const expressionStatementType = getTypeOfNode(
        statement.expression,
        context
      );
      return (
        expressionStatementType !== null &&
        tsApiUtils?.isIntrinsicNeverType(expressionStatementType) === true
      );
    }
    return false;
  };
}

/**
 * Is the given statement, when inside an if statement, a returning branch?
 */
function isIfReturningBranch(statement: TSESTree.Statement) {
  return (
    // Another instance of this rule will check nested if statements.
    isIfStatement(statement) ||
    isReturnStatement(statement) ||
    isThrowStatement(statement) ||
    isBreakStatement(statement) ||
    isContinueStatement(statement)
  );
}

/**
 * Get all of the violations in the given if statement assuming if statements
 * are allowed.
 */
function getIfBranchViolations(
  node: TSESTree.IfStatement,
  context: TSESLint.RuleContext<keyof typeof errorMessages, Options>
): RuleResult<keyof typeof errorMessages, Options>["descriptors"] {
  const branches = [node.consequent, node.alternate];
  const violations = branches.filter<NonNullable<(typeof branches)[0]>>(
    (branch): branch is NonNullable<typeof branch> => {
      if (branch === null || isIfReturningBranch(branch)) {
        return false;
      }

      if (isExpressionStatement(branch)) {
        const expressionStatementType = getTypeOfNode(
          branch.expression,
          context
        );

        if (
          expressionStatementType !== null &&
          tsApiUtils?.isIntrinsicNeverType(expressionStatementType) === true
        ) {
          return false;
        }
      }

      if (isBlockStatement(branch)) {
        if (branch.body.some(isIfReturningBranch)) {
          return false;
        }

        const isNeverExpressions = getIsNeverExpressions(context);
        if (branch.body.some(isNeverExpressions)) {
          return false;
        }
      }

      return true;
    }
  );

  return violations.flatMap(incompleteBranchViolation);
}

/**
 * Is the given statement, when inside a switch statement, a returning branch?
 */
function isSwitchReturningBranch(statement: TSESTree.Statement) {
  return (
    // Another instance of this rule will check nested switch statements.
    isSwitchStatement(statement) ||
    isReturnStatement(statement) ||
    isThrowStatement(statement)
  );
}

/**
 * Get all of the violations in the given switch statement assuming switch
 * statements are allowed.
 */
function getSwitchViolations(
  node: TSESTree.SwitchStatement,
  context: TSESLint.RuleContext<keyof typeof errorMessages, Options>
): RuleResult<keyof typeof errorMessages, Options>["descriptors"] {
  const isNeverExpressions = getIsNeverExpressions(context);

  const violations = node.cases.filter((branch) => {
    if (branch.consequent.length === 0) {
      return false;
    }
    if (branch.consequent.some(isSwitchReturningBranch)) {
      return false;
    }

    if (branch.consequent.every(isBlockStatement)) {
      const lastBlock = branch.consequent.at(-1)!;

      if (lastBlock.body.some(isSwitchReturningBranch)) {
        return false;
      }

      if (lastBlock.body.some(isNeverExpressions)) {
        return false;
      }
    }

    return !branch.consequent.some(isNeverExpressions);
  });

  return violations.flatMap(incompleteBranchViolation);
}

/**
 * Does the given if statement violate this rule if it must be exhaustive.
 */
function isExhaustiveIfViolation(node: TSESTree.IfStatement): boolean {
  return node.alternate === null;
}

/**
 * Does the given typed switch statement violate this rule if it must be exhaustive.
 */
function isExhaustiveTypeSwitchViolation(
  node: TSESTree.SwitchStatement,
  context: TSESLint.RuleContext<keyof typeof errorMessages, Options>
): boolean {
  const discriminantType = getTypeOfNode(node.discriminant, context);
  if (discriminantType === null || !discriminantType.isUnion()) {
    return true;
  }

  const caseTypes = node.cases.reduce<ReadonlySet<Type>>(
    (types, c) => new Set([...types, getTypeOfNode(c.test!, context)!]),
    new Set()
  );

  return discriminantType.types.some((unionType) => !caseTypes.has(unionType));
}

/**
 * Does the given switch statement violate this rule if it must be exhaustive.
 */
function isExhaustiveSwitchViolation(
  node: TSESTree.SwitchStatement,
  context: TSESLint.RuleContext<keyof typeof errorMessages, Options>
): boolean {
  return (
    // No cases defined.
    node.cases.every((c) => c.test !== null)
      ? isExhaustiveTypeSwitchViolation(node, context)
      : false
  );
}

/**
 * Check if the given IfStatement violates this rule.
 */
function checkIfStatement(
  node: TSESTree.IfStatement,
  context: TSESLint.RuleContext<keyof typeof errorMessages, Options>,
  options: Options
): RuleResult<keyof typeof errorMessages, Options> {
  const [{ allowReturningBranches }] = options;

  return {
    context,
    descriptors:
      allowReturningBranches === false
        ? [{ node, messageId: "unexpectedIf" }]
        : allowReturningBranches === "ifExhaustive"
        ? isExhaustiveIfViolation(node)
          ? [{ node, messageId: "incompleteIf" }]
          : getIfBranchViolations(node, context)
        : getIfBranchViolations(node, context),
  };
}

/**
 * Check if the given SwitchStatement violates this rule.
 */
function checkSwitchStatement(
  node: TSESTree.SwitchStatement,
  context: TSESLint.RuleContext<keyof typeof errorMessages, Options>,
  options: Options
): RuleResult<keyof typeof errorMessages, Options> {
  const [{ allowReturningBranches }] = options;

  return {
    context,
    descriptors:
      allowReturningBranches === false
        ? [{ node, messageId: "unexpectedSwitch" }]
        : allowReturningBranches === "ifExhaustive"
        ? isExhaustiveSwitchViolation(node, context)
          ? [{ node, messageId: "incompleteSwitch" }]
          : getSwitchViolations(node, context)
        : getSwitchViolations(node, context),
  };
}

// Create the rule.
export const rule = createRule<keyof typeof errorMessages, Options>(
  name,
  meta,
  defaultOptions,
  {
    IfStatement: checkIfStatement,
    SwitchStatement: checkSwitchStatement,
  }
);
