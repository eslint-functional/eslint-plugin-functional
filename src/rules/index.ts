import * as functionalParameters from "./functional-parameters";
import * as immutableData from "./immutable-data";
import * as noClass from "./no-class";
import * as noConditionalStatement from "./no-conditional-statement";
import * as noExpressionStatement from "./no-expression-statement";
import * as noLet from "./no-let";
import * as noLoop from "./no-loop-statement";
import * as noMethodSignature from "./no-method-signature";
import * as noMixedType from "./no-mixed-type";
import * as noPromiseReject from "./no-promise-reject";
import * as noReturnVoid from "./no-return-void";
import * as noThisExpression from "./no-this-expression";
import * as noThrowStatement from "./no-throw-statement";
import * as noTryStatement from "./no-try-statement";
import * as preferReadonlyTypes from "./prefer-readonly-type";
import * as preferType from "./prefer-type-literal";

/**
 * All of the custom rules.
 */
export const rules = {
  [functionalParameters.name]: functionalParameters.rule,
  [immutableData.name]: immutableData.rule,
  [noClass.name]: noClass.rule,
  [noConditionalStatement.name]: noConditionalStatement.rule,
  [noExpressionStatement.name]: noExpressionStatement.rule,
  [noLet.name]: noLet.rule,
  [noLoop.name]: noLoop.rule,
  [noMethodSignature.name]: noMethodSignature.rule,
  [noMixedType.name]: noMixedType.rule,
  [noPromiseReject.name]: noPromiseReject.rule,
  [noReturnVoid.name]: noReturnVoid.rule,
  [noThisExpression.name]: noThisExpression.rule,
  [noThrowStatement.name]: noThrowStatement.rule,
  [noTryStatement.name]: noTryStatement.rule,
  [preferReadonlyTypes.name]: preferReadonlyTypes.rule,
  [preferType.name]: preferType.rule,
};
