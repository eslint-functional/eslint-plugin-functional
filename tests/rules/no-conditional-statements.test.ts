import dedent from "dedent";
import { createRuleTester } from "eslint-vitest-rule-tester";
import { describe, expect, it } from "vitest";

import { name, rule } from "#/rules/no-conditional-statements";

import { typescriptConfig } from "../utils/configs";

describe(name, () => {
  describe("typescript", () => {
    const { valid, invalid } = createRuleTester({
      name,
      rule,
      configs: typescriptConfig,
    });

    describe("if statements", () => {
      it("reports if statements", async () => {
        const invalidResult = await invalid({
          code: dedent`
            if (true) {
              console.log("hello world");
            }
          `,
          errors: ["unexpectedIf"],
        });
        expect(invalidResult.result).toMatchSnapshot();
      });

      describe("options", () => {
        describe("allowReturningBranches", () => {
          describe("true", () => {
            it("allows return", async () => {
              await valid({
                code: dedent`
                  function foo(i) {
                    if (i === 1) {
                      return 1;
                    }
                  }
                `,
                options: [{ allowReturningBranches: true }],
              });
            });

            it("supports break and continue as returning", async () => {
              await valid({
                code: dedent`
                  for(var i = 0; i < j; i++) {
                    if (e === 1) {
                      break;
                    }
                    if (e === 2) {
                      continue;
                    }
                  }
                `,
                options: [{ allowReturningBranches: true }],
              });
            });

            it("supports throw as returning", async () => {
              await valid({
                code: dedent`
                  function foo(i) {
                    if (i === 1) {
                      throw 1;
                    }
                  }
                `,
                options: [{ allowReturningBranches: true }],
              });
            });

            it("supports never as returning", async () => {
              await valid({
                code: dedent`
                  declare function neverReturn(): never;
                  function foo(i) {
                    if (i === 1) {
                      neverReturn();
                    }
                  }
                `,
                options: [{ allowReturningBranches: true }],
              });
            });
          });

          describe("ifExhaustive", () => {
            it("else required", async () => {
              await valid({
                code: dedent`
                  function foo(i) {
                    if (i === 1) {
                      return 1;
                    } else {
                      return 0;
                    }
                  }
                `,
                options: [{ allowReturningBranches: "ifExhaustive" }],
              });

              const invalidResult = await invalid({
                code: dedent`
                  function foo(i) {
                    if (i === 1) {
                      return 1;
                    }
                  }
                `,
                options: [{ allowReturningBranches: "ifExhaustive" }],
                errors: ["incompleteIf"],
              });
              expect(invalidResult.result).toMatchSnapshot();
            });

            it("supports break and continue as returning", async () => {
              await valid({
                code: dedent`
                  for(var i = 0; i < j; i++) {
                    if (e === 1) {
                      break;
                    } else {
                      continue;
                    }
                  }
                `,
                options: [{ allowReturningBranches: "ifExhaustive" }],
              });
            });
          });
        });

        describe("ignoreCodePattern", () => {
          it("ignores matching conditionals", async () => {
            await valid({
              code: dedent`
                if (import.meta.vitest) {
                  const { it, expect } = import.meta.vitest;
                }
              `,
              options: [{ ignoreCodePattern: ["import\\.meta\\.vitest"] }],
            });
          });
        });
      });
    });

    describe("switch statements", () => {
      it("reports switch statements", async () => {
        const invalidResult = await invalid({
          code: dedent`
            switch(i) {
              case "a":
                console.log("hello");
              case "b":
              case "c":
                console.log("world");
            }
          `,
          errors: ["unexpectedSwitch"],
        });
        expect(invalidResult.result).toMatchSnapshot();
      });

      describe("options", () => {
        describe("allowReturningBranches", () => {
          describe("true", () => {
            it("allows return", async () => {
              await valid({
                code: dedent`
                  function foo(i) {
                    switch(i) {
                      case "a":
                        return 1;
                      case "b":
                        return 2;
                      default:
                        return 3;
                    }
                  }
                `,
                options: [{ allowReturningBranches: true }],
              });
            });

            it("supports break and continue as returning", async () => {
              await valid({
                code: dedent`
                  label: for(var i = 0; i < j; i++) {
                    switch(i) {
                      case "a":
                        break label;
                      case "b":
                        continue;
                    }
                  }
                `,
                options: [{ allowReturningBranches: true }],
              });
            });

            it("supports throw as returning", async () => {
              await valid({
                code: dedent`
                  function foo(i) {
                    switch(i) {
                      case "a":
                        throw 1;
                      case "b":
                        throw 2;
                    }
                  }
                `,
                options: [{ allowReturningBranches: true }],
              });
            });

            it("supports never as returning", async () => {
              await valid({
                code: dedent`
                  declare function neverReturn(): never;
                  function foo(i) {
                    switch(i) {
                      case "a":
                        neverReturn();
                      case "b":
                        neverReturn();
                    }
                  }
                `,
                options: [{ allowReturningBranches: true }],
              });
            });
          });

          describe("ifExhaustive", () => {
            it("requires default case", async () => {
              await valid({
                code: dedent`
                  function foo(i) {
                    switch(i) {
                      case "a":
                        return 1;
                      case "b":
                        return 2;
                      default:
                        return 3;
                    }
                  }
                `,
                options: [{ allowReturningBranches: "ifExhaustive" }],
              });

              const invalidResult = await invalid({
                code: dedent`
                  function foo(i) {
                    switch(i) {
                      case "a":
                        return 1;
                      case "b":
                        return 2;
                    }
                  }
                `,
                options: [{ allowReturningBranches: "ifExhaustive" }],
                errors: ["incompleteSwitch"],
              });
              expect(invalidResult.result).toMatchSnapshot();
            });

            it("supports exhaustive type testing", async () => {
              await valid({
                code: dedent`
                  type T = "a" | "b";
                  function foo(i: T) {
                    switch(i) {
                      case "a":
                        return 1;
                      case "b":
                        return 2;
                    }
                  }
                `,
                options: [{ allowReturningBranches: "ifExhaustive" }],
              });
            });
          });
        });
      });
    });
  });
});
