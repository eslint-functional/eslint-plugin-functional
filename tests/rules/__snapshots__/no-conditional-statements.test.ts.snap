// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`no-conditional-statements > typescript > if statements > options > allowReturningBranches > ifExhaustive > else required 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 3,
      "endColumn": 4,
      "endLine": 4,
      "line": 2,
      "message": "Incomplete if, it must have an else statement and every branch must contain a return statement.",
      "messageId": "incompleteIf",
      "nodeType": "IfStatement",
      "ruleId": "no-conditional-statements",
      "severity": 2,
    },
  ],
  "output": "function foo(i) {
  if (i === 1) {
    return 1;
  }
}",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 3,
          "endColumn": 4,
          "endLine": 4,
          "line": 2,
          "message": "Incomplete if, it must have an else statement and every branch must contain a return statement.",
          "messageId": "incompleteIf",
          "nodeType": "IfStatement",
          "ruleId": "rule-to-test/no-conditional-statements",
          "severity": 2,
        },
      ],
      "output": "function foo(i) {
  if (i === 1) {
    return 1;
  }
}",
    },
  ],
}
`;

exports[`no-conditional-statements > typescript > if statements > reports if statements 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 1,
      "endColumn": 2,
      "endLine": 3,
      "line": 1,
      "message": "Unexpected if, use a conditional expression (ternary operator) instead.",
      "messageId": "unexpectedIf",
      "nodeType": "IfStatement",
      "ruleId": "no-conditional-statements",
      "severity": 2,
    },
  ],
  "output": "if (true) {
  console.log("hello world");
}",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 1,
          "endColumn": 2,
          "endLine": 3,
          "line": 1,
          "message": "Unexpected if, use a conditional expression (ternary operator) instead.",
          "messageId": "unexpectedIf",
          "nodeType": "IfStatement",
          "ruleId": "rule-to-test/no-conditional-statements",
          "severity": 2,
        },
      ],
      "output": "if (true) {
  console.log("hello world");
}",
    },
  ],
}
`;

exports[`no-conditional-statements > typescript > switch statements > options > allowReturningBranches > ifExhaustive > requires default case 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 3,
      "endColumn": 4,
      "endLine": 7,
      "line": 2,
      "message": "Incomplete switch, it must be exhaustive or have an default case and every case must contain a return statement.",
      "messageId": "incompleteSwitch",
      "nodeType": "SwitchStatement",
      "ruleId": "no-conditional-statements",
      "severity": 2,
    },
  ],
  "output": "function foo(i) {
  switch(i) {
    case "a":
      return 1;
    case "b":
      return 2;
  }
}",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 3,
          "endColumn": 4,
          "endLine": 7,
          "line": 2,
          "message": "Incomplete switch, it must be exhaustive or have an default case and every case must contain a return statement.",
          "messageId": "incompleteSwitch",
          "nodeType": "SwitchStatement",
          "ruleId": "rule-to-test/no-conditional-statements",
          "severity": 2,
        },
      ],
      "output": "function foo(i) {
  switch(i) {
    case "a":
      return 1;
    case "b":
      return 2;
  }
}",
    },
  ],
}
`;

exports[`no-conditional-statements > typescript > switch statements > reports switch statements 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 1,
      "endColumn": 2,
      "endLine": 7,
      "line": 1,
      "message": "Unexpected switch, use a conditional expression (ternary operator) instead.",
      "messageId": "unexpectedSwitch",
      "nodeType": "SwitchStatement",
      "ruleId": "no-conditional-statements",
      "severity": 2,
    },
  ],
  "output": "switch(i) {
  case "a":
    console.log("hello");
  case "b":
  case "c":
    console.log("world");
}",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 1,
          "endColumn": 2,
          "endLine": 7,
          "line": 1,
          "message": "Unexpected switch, use a conditional expression (ternary operator) instead.",
          "messageId": "unexpectedSwitch",
          "nodeType": "SwitchStatement",
          "ruleId": "rule-to-test/no-conditional-statements",
          "severity": 2,
        },
      ],
      "output": "switch(i) {
  case "a":
    console.log("hello");
  case "b":
  case "c":
    console.log("world");
}",
    },
  ],
}
`;
